test/gtest/common/test_helpers.h

```diff
@@ -1053,6 +1054,77 @@ const std::vector<std::vector<ucs_memory_type_t> >& supported_mem_type_pairs();
  */
 void skip_on_address_sanitizer();
 
+/**
+ * Class for mocking C function pointers
+ * The class is used to mock C function pointers, by replacing the original
+ * function pointer with a mock function pointer. The original function pointer
+ * is restored when the object is destroyed.
+ */
+class mock {
+public:
+    using func = void (*)();
+
+    ~mock()
+    {
+        cleanup();
+    }
+
+    void cleanup()
+    {
+        m_mock_map.clear();
+    }
+
+    template <typename Fn> void setup(Fn *orig_ptr, Fn mock)
+    {
+        func *ptr    = reinterpret_cast<func*>(orig_ptr);
+        auto it      = m_mock_map.find(ptr);
+        if (it == m_mock_map.end()) {
+            mock_func *new_mock = new mock_func(ptr, (func)mock);
+            m_mock_map[ptr]     = std::unique_ptr<mock_func>(new_mock);
+        } else {
+            it->second->m_mock = (func)mock;
```

Comment:
don't we want to replace *orig_ptr here?
do we have to save m_mock in mock_func as a member var?